src/rust/engine/src/core.rs

```diff
@@ -114,18 +116,32 @@ impl fmt::Display for Params {
 
 pub type Id = u64;
 
-// The type of a python object (which itself has a type, but which is not represented
-// by a Key, because that would result in a infinitely recursive structure.)
-#[repr(C)]
+// A pointer to an underlying PyTypeObject instance.
 #[derive(Clone, Copy, Eq, Hash, Ord, PartialEq, PartialOrd)]
-pub struct TypeId(pub Id);
+pub struct TypeId(*mut std::ffi::c_void);
```

Comment:
Maybe this isn't necessary if others have more C experience, but consider a comment like this:

> Note that we use C's `*void` pointer because we don't know what the type may be.

Either way, things I (re)learned.